test/test_app.py

```diff
@@ -0,0 +1,33 @@
+import pytest
+
+from flask.testing import FlaskClient
+
+from flask_calendar.app import app
+from flask_calendar.constants import SESSION_ID
+
+app.testing = True
+
+
+@pytest.fixture
+def client() -> FlaskClient:
+    return app.test_client()
+
+
+@pytest.mark.parametrize(('username', 'password', 'success'), (
+    ('', '', False),
+    ('wrong_username', '', False),
+    ('a_username', '', False),
+    ('a_username', 'wrong_password', False),
+    ('a_username', 'a_password', True)
+))
+def test_login_credentials(client: FlaskClient, username: str, password: str, success: bool) -> None:
+    rv = client.post('/do_login', data=dict(
+        username=username,
+        password=password
+    ), follow_redirects=True)
+    cookie = next((c for c in client.cookie_jar), None)
+    assert rv.status_code == 200
```

Comment:
If you want to be purist, this sounds like two tests:
- one to check responses of login attempts (`test_login_credentials` sounds good)
- another to check cookie(s) once logged in (`test_session_id_cookie_present_when_logged_in` or similar)

I don't always follow this rule so just take it as a suggestion for you to decide, not a requirement.

Reply:
Having two tests, one for credentials and one for the session cookie, seems the way to go. This forces me to check if the credentials were valid or not in a different way.